golem-worker-executor-base/src/storage/cassandra.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use futures::StreamExt;
+use golem_common::config::CassandraConfig;
+use golem_common::metrics::db::{record_db_failure, record_db_success};
+use scylla::batch::{Batch, BatchType};
+use scylla::prepared_statement::PreparedStatement;
+use scylla::query::Query;
+use scylla::serialize::row::SerializeRow;
+use scylla::transport::errors::QueryError;
+use scylla::FromRow;
+use scylla::{transport::session::PoolSize, Session, SessionBuilder};
+use std::fmt::Debug;
+use std::time::Instant;
+use std::{num::NonZeroUsize, sync::Arc};
+
+#[derive(Debug, Clone)]
+pub struct CassandraSession {
+    pub session: Arc<Session>,
+    pub keyspace: String,
+    pub set_tracing: bool,
+}
+
+impl CassandraSession {
+    pub fn new(session: Arc<Session>, set_tracing: bool, keyspace: &str) -> Self {
+        CassandraSession {
+            session,
+            keyspace: String::from(keyspace),
+            set_tracing,
+        }
+    }
+
+    pub async fn configured(config: &CassandraConfig) -> Result<Self, String> {
+        let mut session_builder = SessionBuilder::new()
+            .known_nodes_addr(config.nodes.iter())
+            .pool_size(PoolSize::PerHost(
+                NonZeroUsize::new(config.pool_size_per_host).unwrap(),
+            ))
+            .use_keyspace(&config.keyspace, false);
+
+        if let (Some(username), Some(password)) =
+            (config.username.as_ref(), config.password.as_ref())
+        {
+            session_builder = session_builder.user(username, password);
+        }
+
+        let session = session_builder.build().await.map_err(|e| e.to_string())?;
+
+        Ok(CassandraSession {
+            session: Arc::new(session),
+            keyspace: config.keyspace.clone(),
+            set_tracing: config.tracing,
+        })
+    }
+
+    pub async fn create_schema(&self) -> Result<(), String> {
```

Comment:
i would still move the (test) schema creations out of the common cassandra (and sqlite) code, and even separate them by usage (for sqlite), but if we create an issue for it, then i'm okay with it for now

Reply:
I think there is only one common schema for the storage, to where you want to move the schema creation ? , 
how you want to separate them ?

Reply:
based on the storage type: kv, blob, indexed; and move to that package.
with ideally we can select separately what to use, and only install what is needed (even if it is only for tests)

Reply:
I don't think it makes sense, because I don't think we will never use  kv storage with sqlite and indexed storage in Cassandra or some approach like that, and I don't see neither any advantage in performance or space usage or something useful, but I did it because I think we need to move on with this PR.

Reply:
The three storage "types" (kv, indexed, blob) should be completely separate and separately configurable. Even though sqlite + cassandra for example is not a likely combination, it is possible that we will have more backend implementations in the future that make sense to combine them in ways we don't see right now. In general, using something different for kv-store (basically caching data) and indexed-store (our primary storage layer for durable execution) completely makes sense to me.